File: whitelabel-copytrade/apps/admin-web/src/modules/billing/portal/billing-portal-api.ts
/**
 * Billing Portal API client for admin-web.
 * Consumes canonical billing APIs, never hardcodes pricing.
 * No secrets exposed.
 */

const API_BASE = '/api/v1/billing/portal';

async function fetchJson(url: string, options?: RequestInit): Promise<any> {
  const res = await fetch(url, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...(options?.headers || {}),
    },
    credentials: 'include',
  });
  if (!res.ok) {
    const errText = await res.text().catch(() => '');
    throw new Error(errText || `Request failed: ${res.status}`);
  }
  return res.json();
}

export async function getBillingOverview(): Promise<any> {
  return fetchJson(`${API_BASE}/overview`);
}

export async function getCurrentSubscription(): Promise<any> {
  return fetchJson(`${API_BASE}/subscription`);
}

export async function getAvailablePlans(): Promise<any> {
  return fetchJson(`${API_BASE}/plans`);
}

export async function getPlanComparison(): Promise<any> {
  return fetchJson(`${API_BASE}/plans/comparison`);
}

export async function getUsageSummary(): Promise<any> {
  return fetchJson(`${API_BASE}/usage`);
}

export async function listInvoices(params?: { status?: string; limit?: number; fromDate?: string; toDate?: string }): Promise<any> {
  const q = new URLSearchParams();
  if (params?.status) q.append('status', params.status);
  if (params?.limit) q.append('limit', String(params.limit));
  if (params?.fromDate) q.append('fromDate', params.fromDate);
  if (params?.toDate) q.append('toDate', params.toDate);
  return fetchJson(`${API_BASE}/invoices?${q.toString()}`);
}

export async function getInvoiceDetail(id: string): Promise<any> {
  return fetchJson(`${API_BASE}/invoices/${id}`);
}

export async function getInvoicePdfMetadata(id: string): Promise<any> {
  return fetchJson(`${API_BASE}/invoices/${id}/pdf-metadata`);
}

export async function listPayments(params?: { status?: string; provider?: string; limit?: number }): Promise<any> {
  const q = new URLSearchParams();
  if (params?.status) q.append('status', params.status);
  if (params?.provider) q.append('provider', params.provider);
  if (params?.limit) q.append('limit', String(params.limit));
  return fetchJson(`${API_BASE}/payments?${q.toString()}`);
}

export async function getPaymentDetail(id: string): Promise<any> {
  return fetchJson(`${API_BASE}/payments/${id}`);
}

export async function getPaymentStatus(id: string): Promise<any> {
  return fetchJson(`${API_BASE}/payments/${id}/status`);
}

export async function createCheckoutSession(payload: { planId: string; billingInterval?: string; currency?: string; provider?: string; successUrl?: string; cancelUrl?: string; idempotencyKey?: string }): Promise<any> {
  return fetchJson(`${API_BASE}/checkout`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function getCheckoutStatus(id: string): Promise<any> {
  return fetchJson(`${API_BASE}/checkout/${id}/status`);
}

export async function cancelSubscription(payload: { reason?: string; atPeriodEnd?: boolean }): Promise<any> {
  return fetchJson(`${API_BASE}/subscription/cancel`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function resumeSubscription(): Promise<any> {
  return fetchJson(`${API_BASE}/subscription/resume`, {
    method: 'POST',
    body: JSON.stringify({}),
  });
}

export async function changePlan(payload: { planId: string; atPeriodEnd?: boolean }): Promise<any> {
  return fetchJson(`${API_BASE}/subscription/change-plan`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function changeInterval(payload: { newInterval: string; atPeriodEnd?: boolean }): Promise<any> {
  return fetchJson(`${API_BASE}/subscription/change-interval`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

File: whitelabel-copytrade/apps/mobile/lib/features/billing/portal/billing_portal_api.dart
/// Billing Portal API client for mobile - consumes canonical APIs, no hardcoded pricing.

import '../../../core/network/api_client.dart';

class BillingPortalApi {
  final ApiClient _client;

  BillingPortalApi(this._client);

  Future<Map<String, dynamic>> getBillingOverview() async {
    final res = await _client.get('/v1/billing/portal/overview');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getCurrentSubscription() async {
    final res = await _client.get('/v1/billing/portal/subscription');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getAvailablePlans() async {
    final res = await _client.get('/v1/billing/portal/plans');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getPlanComparison() async {
    final res = await _client.get('/v1/billing/portal/plans/comparison');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getUsageSummary() async {
    final res = await _client.get('/v1/billing/portal/usage');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> listInvoices({String? status, int? limit}) async {
    final query = <String, dynamic>{};
    if (status != null) query['status'] = status;
    if (limit != null) query['limit'] = limit;
    final res = await _client.get('/v1/billing/portal/invoices', queryParameters: query);
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getInvoiceDetail(String id) async {
    final res = await _client.get('/v1/billing/portal/invoices/$id');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> listPayments({String? status, String? provider, int? limit}) async {
    final query = <String, dynamic>{};
    if (status != null) query['status'] = status;
    if (provider != null) query['provider'] = provider;
    if (limit != null) query['limit'] = limit;
    final res = await _client.get('/v1/billing/portal/payments', queryParameters: query);
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getPaymentStatus(String id) async {
    final res = await _client.get('/v1/billing/portal/payments/$id/status');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> createCheckoutSession({
    required String planId,
    String? billingInterval,
    String? currency,
    String? provider,
    String? successUrl,
    String? cancelUrl,
    String? idempotencyKey,
  }) async {
    final res = await _client.post('/v1/billing/portal/checkout', data: {
      'planId': planId,
      if (billingInterval != null) 'billingInterval': billingInterval,
      if (currency != null) 'currency': currency,
      if (provider != null) 'provider': provider,
      if (successUrl != null) 'successUrl': successUrl,
      if (cancelUrl != null) 'cancelUrl': cancelUrl,
      if (idempotencyKey != null) 'idempotencyKey': idempotencyKey,
    });
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getCheckoutStatus(String id) async {
    final res = await _client.get('/v1/billing/portal/checkout/$id/status');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> cancelSubscription({String? reason, bool? atPeriodEnd}) async {
    final res = await _client.post('/v1/billing/portal/subscription/cancel', data: {
      if (reason != null) 'reason': reason,
      if (atPeriodEnd != null) 'atPeriodEnd': atPeriodEnd,
    });
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> resumeSubscription() async {
    final res = await _client.post('/v1/billing/portal/subscription/resume');
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> changePlan({required String planId, bool? atPeriodEnd}) async {
    final res = await _client.post('/v1/billing/portal/subscription/change-plan', data: {
      'planId': planId,
      if (atPeriodEnd != null) 'atPeriodEnd': atPeriodEnd,
    });
    return res.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> changeInterval({required String newInterval, bool? atPeriodEnd}) async {
    final res = await _client.post('/v1/billing/portal/subscription/change-interval', data: {
      'newInterval': newInterval,
      if (atPeriodEnd != null) 'atPeriodEnd': atPeriodEnd,
    });
    return res.data as Map<String, dynamic>;
  }
}
